### PR TITLE
fix bug where if maxiter=0 with lsq-exact optimizer, it would never t…

### PR DESCRIPTION
**Pull request by dpanici**
- fix bug where if maxiter=0 with lsq-exact optimizer, it would never terminate since check_termination would never get called as nfev>max_nfev from the start
- fix bug where if a desc bfgs method is used, the xscale would still default to `hess` and then the optimizer throws an error
- passing a string of the name of an optimizer to `eq.solve` now will actually use the requested optimizer
- fix bug where when using `sgd`, `method` is passed instead of `self.method`
- add tests for maxiter=0 and maxiter=1 for desc least squares and scalar

### CODE REVIEW

**Comment by f0uriest** [comment]
I thought we fixed this a while ago? Both ``success`` and ``message`` get initialized, so we don't have to call ``check_termination``

**Comment by dpanici** [comment]
```
import desc.examples
eq = desc.examples.get('SOLOVEV')
eq.solve(maxiter=0,verbose=3)
```
yields this on master

image

**Comment by dpanici** [comment]
it was fixed for #230 nfev=1 but not maxiter=0

**Comment by f0uriest** [comment]
ah ok we should check if `fmin_scalar` is also affected, we might need to fix that as well

**Review by ddudt — CHANGES_REQUESTED** [review]
Looks good, but we should include small tests for these edge cases.

File: desc/optimize/least_squares.py [hunk 1]
```diff
@@ -221,6 +221,27 @@ def lsqtr(  # noqa: C901 - FIXME: simplify this
 
         actual_reduction = -1
 
+        if max_nfev or maxiter == 0:
```
Comment (f0uriest) [inline desc/optimize/least_squares.py:224]: do we need this if? or can we always just check termination before starting the loop to see if we're already at a local min?

File: desc/optimize/optimizer.py [hunk 2]
```diff
@@ -494,7 +498,7 @@ def jac(x_reduced):
                 x0=x0_reduced,
                 grad=grad_wrapped,
                 args=(),
-                method=method,
+                method=self.method,
```
Comment (f0uriest) [inline desc/optimize/optimizer.py:501]: i think this needs to be method since we parse the string above to detect some options

**Review by f0uriest — CHANGES_REQUESTED** [review]
Did we also check that `fmintr` is not affected by this same bug?

File: desc/optimize/optimizer.py (hunk 2 above)
Comment (dpanici) [reply desc/optimize/optimizer.py:501]: the parse only happens for the elif statement above, not inside this elif statement, so method is undefined here. also currently no options affect the sgd method

File: desc/optimize/least_squares.py (hunk 1 above)
Comment (dpanici) [reply desc/optimize/least_squares.py:224]: a few of the values needed for check_termination are not initially defined until the while loop, and the way the code is structured, it always check for termination since it will always enter the loop once. this just catches the case where with 0 maxiter, it never enters the loop and so never checks for termination

**Comment by dpanici** [comment]
> Did we also check that `fmintr` is not affected by this same bug?

yep, the test with `dogleg-bfgs`  method tests fmintr

**Review by ddudt — APPROVED** [review]
